shared-libs/transitions/src/transitions/registration.js

```diff
@@ -125,6 +126,49 @@ const getDaysSinceDOB = doc => {
   return findFirstDefinedValue(doc, fields);
 };
 
+const getLMPDateFromParts = doc => {
+  //basic validations will be done in the app_settings.registrations
+  return getDateFromParts(doc.fields.lmpYYYY, doc.fields.lmpMM, doc.fields.lmpDD);
+};
+
+const getLMPDateFromString = dateString => {
+  const separator = dateString[dateString.search(/[^0-9]/)];//Find non-numeric character
+  const dateParts = dateString.split(separator);
+  return getDateFromParts(dateParts[0], dateParts[1], dateParts[2]);
+};
+
+const getDateFromParts = (year, month, day) => {
+
+  const dateYYYY = year.toString().padStart(4, '2000');
+  const dateMM = month.toString().padStart(2, '00');
+  const dateDD = day.toString().padStart(2, '00');
+
+  let gregDateISO;
+
+  //Bikram Sambat is either 56 or 57 years ahead of Gregorian
+  //To support LMP dates from last year also, we check now + 55 only
+  if (dateYYYY > moment().year() + 55) {
```

Comment:
It'd be nice to have in `bikram-sambat` library a function that can tell if a date is Bikram Sambar date, it can be reused in other places. 

Related with the limitation # 2:

> Bikram Sambat calendar is always either 56 or 57 years ahead of Gregorian. If the specified year is more than the current Gregorian year + 55 (e.g. 2021 + 55 = 2076), then it is assumed to be in Bikram Sambat and will be converted to Gregorian. Another option would be to specify the preferred calendar explicitly, but that might not be necessary because we only accept dates in the recent past anyway 

Maybe a form's field or project configuration can tell the calendar to use. 
How are dates currently working in other forms or places in the app?

Reply:
Thanks, @latin-panda 
Just looking at the date, it is not always immediately clear whether it is in Bikram Sambat or Gregorian.
For example, today's Gregorian date "2021-12-15" is also a valid date in Bikram Sambat, though it was 57 years ago.
For the purpose of the LMP date for current pregnancy registration, we can safely assume that this date is not in Bikram Sambat. However, for other dates such as DOB, this assumption does not work.

We can still add support for this function in the `bikram-sambat` library where we also provide the number of years in the past to consider. It should also be able to tell dates such as "2021-02-29" which is not a valid date in Gregorian but can be valid in Bikram Sambat.

Yes, it would be great to have a configuration switch to specify the calendar system. I looked at the `smsParser` which already supports [`date` ](https://github.com/medic/cht-core/blob/1b887c5bbe2957ebbea6246ebbed824ecb080060/api/src/services/report/smsparser.js#L156) type. I tried adding `bsDate` but it applied to a single field. Our primary aim is to support space between date parts e.g. "B 12345 2078 03 03". (Type B in my description above).

Currently, in the app forms, we can get a Bikram Sambat date picker using `bikram-sambat` in the appearance column.
With [this](https://github.com/medic/cht-core/issues/7294), we can use Bikram Sambat dates throughout the app when Nepali locale is selected.

Reply:
Thanks for the explanation Binod! It's very helpful 

As I can see in this code, we need to parse from `Bikram Sambat` date to `Gregorian` so that the `Registration` transition can do its original work, please correct me if I'm missing something :) 

Looking at the code, it'd be easier if:

**Op 1:**  
Create a new field type where you pointed before in the [smsParser](https://github.com/medic/cht-core/blob/1b887c5bbe2957ebbea6246ebbed824ecb080060/api/src/services/report/smsparser.js#L156). Here we can make the conversion to `Gregorian` assuming the date are coming as `Bikram Sambat`, to keep it simple. 
Then we can configure the forms to use this new field and add validations rules to send an sms with the error if the date is invalid:
```
"validations": {
      "list": [
        {
          "property": "lmp_bkdate",
          "rule": ".....[your regex or 8 weeks and 40 weeks validations]....",
          "translation_key": "messages.hr.validation.lmp_invalid_format"
        }
```
This solution flows with the current system, will give more flexibility to future implementations and resolve the limitations about not notifying the user when failing in the transition here. But we can't use spaces between YYYY MM DD.

**Op 2**: 
Get and use the locale ([available in forms](https://github.com/medic/cht-core/blob/master/api/src/services/records.js#L175)) to check if it's using Nepal and then do the dates conversions to Gregorian. 

I think using spaces when entering the date is adding complexity, like adding more fixed variables that we can't rename or add levels, the 8/40 weeks validations is happening in the transition where it's harder to notify back to the user with sms (I think 🤔)

Is it possible to drop the spaces? Asking since I'm not aware yet of the full picture of the user's need

Note: _I hope this is helpful, I'll be out the next 2 weeks but please do reach out if anything, another teammate might help or me when I'm back_  :)

Reply:
Using spaces as the delimiter is the preferred method because the CHWs in Nepal use it all the time when sending other SMS forms. It can be difficult for many CHWs to learn to type another symbol such as `-` or `/`. Almost all of them use simple keypad phones. Still, we also want to support regular formats such as `2021-12-15 and 2021/12/15.

Here are some relevant docs on how we decided on the format/design (might be inaccessible to public):
- [Nepal MoH - Using Date in LMP: Design Considerations ](https://docs.google.com/document/d/1HoYodpoHXUU1RsQ05U2wMH3lXrUSp8x3h0ibTy_Q4BU/edit#heading=h.iaoayk6ew2xx)
- [LMP Date test findings](https://docs.google.com/presentation/d/1bkDC69BWm0KELLSAV48r_3oR5qpqk43Bj_kEIjKbvcw/edit#slide=id.p)
- [LMP date format design](https://docs.google.com/document/d/1rgTBFlMPKRmbzXcFo-eN3S3AA8d38ChNL0bc1wHEUNw/edit)

Thanks for putting up the options. I will give it some more thinking. Enjoy your holidays @latin-panda.

Reply:
The PR is now ready for review. Most of the work is done in `smsParser` and `validation` now.